File: experiment-5/runs/language=java_model=claude-opus-4-8_tooling=none/rep3/src/test/java/com/brazilsoccer/mcp/McpServerTest.java
/*
 * ============================================================================
 * McpServerTest.java
 * ============================================================================
 * Context:
 *   BDD (Given/When/Then) coverage of the MCP protocol surface: the initialize
 *   handshake, tools/list catalogue, tools/call dispatch and error handling,
 *   plus an end-to-end stdio round-trip. Exercises SoccerTools formatting via
 *   the JSON-RPC layer so the whole server is validated together.
 * ============================================================================
 */
package com.brazilsoccer.mcp;

import com.brazilsoccer.mcp.server.McpServer;
import com.brazilsoccer.mcp.server.SoccerTools;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class McpServerTest {

    private final ObjectMapper json = new ObjectMapper();

    private McpServer server() {
        return new McpServer(new SoccerTools(TestData.store()));
    }

    @Test
    @DisplayName("Given an initialize request, When handled, Then protocol version and capabilities are returned")
    void initializeHandshake() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"initialize\",\"params\":{}}");

        ObjectNode resp = server().handle(req);

        assertEquals(McpServer.PROTOCOL_VERSION, resp.path("result").path("protocolVersion").asText());
        assertTrue(resp.path("result").path("capabilities").has("tools"));
        assertEquals(McpServer.SERVER_NAME, resp.path("result").path("serverInfo").path("name").asText());
    }

    @Test
    @DisplayName("Given a tools/list request, When handled, Then all seven tools are advertised with schemas")
    void toolsListed() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"id\":2,\"method\":\"tools/list\"}");

        ObjectNode resp = server().handle(req);

        JsonNode tools = resp.path("result").path("tools");
        assertTrue(tools.isArray());
        Set<String> names = new HashSet<>();
        for (JsonNode t : tools) {
            names.add(t.path("name").asText());
            assertTrue(t.path("inputSchema").path("type").asText().equals("object"),
                    "each tool must declare an object input schema");
        }
        assertTrue(names.containsAll(Set.of(
                "search_matches", "team_record", "head_to_head", "search_players",
                "competition_standings", "league_stats", "top_scoring_teams")));
    }

    @Test
    @DisplayName("Given a tools/call for standings, When handled, Then the text mentions the champion")
    void toolsCallStandings() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\",\"params\":"
                        + "{\"name\":\"competition_standings\","
                        + "\"arguments\":{\"competition\":\"Brasileirão\",\"season\":2019}}}");

        ObjectNode resp = server().handle(req);

        JsonNode result = resp.path("result");
        assertFalse(result.path("isError").asBoolean());
        String text = result.path("content").get(0).path("text").asText();
        assertTrue(text.contains("Champion"), "standings output should label the champion");
        assertTrue(text.toLowerCase().contains("flamengo"));
    }

    @Test
    @DisplayName("Given a tools/call with a missing required arg, When handled, Then an isError result is returned")
    void toolsCallMissingArg() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"id\":4,\"method\":\"tools/call\",\"params\":"
                        + "{\"name\":\"team_record\",\"arguments\":{}}}");

        ObjectNode resp = server().handle(req);

        assertTrue(resp.path("result").path("isError").asBoolean(),
                "missing required 'team' should produce an isError result");
    }

    @Test
    @DisplayName("Given an unknown method, When handled, Then a JSON-RPC method-not-found error is returned")
    void unknownMethod() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"id\":5,\"method\":\"does/notExist\"}");

        ObjectNode resp = server().handle(req);

        assertEquals(-32601, resp.path("error").path("code").asInt());
    }

    @Test
    @DisplayName("Given a notification, When handled, Then no response is produced")
    void notificationProducesNoResponse() throws Exception {
        JsonNode req = json.readTree(
                "{\"jsonrpc\":\"2.0\",\"method\":\"notifications/initialized\"}");

        ObjectNode resp = server().handle(req);

        assertNull(resp, "notifications must not produce a response");
    }

    @Test
    @DisplayName("Given a stdio stream of requests, When served, Then a JSON response line is emitted per request")
    void stdioRoundTrip() throws Exception {
        String input =
                "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"initialize\",\"params\":{}}\n"
                        + "{\"jsonrpc\":\"2.0\",\"id\":2,\"method\":\"tools/call\",\"params\":"
                        + "{\"name\":\"search_players\",\"arguments\":{\"name\":\"Neymar\"}}}\n";
        ByteArrayInputStream in = new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8));
        ByteArrayOutputStream out = new ByteArrayOutputStream();

        server().serve(in, out);

        String[] lines = out.toString(StandardCharsets.UTF_8).trim().split("\n");
        assertEquals(2, lines.length, "expected one response line per request");
        JsonNode second = json.readTree(lines[1]);
        assertTrue(second.path("result").path("content").get(0).path("text").asText().contains("Neymar"));
    }
}

File: experiments/adrianco/experiment-7/brazil/runs/language=clojure_model=claude-opus-4-8-fast/rep3/test/brazilian_soccer/mcp_server_test.clj
;; =============================================================================
;; brazilian-soccer.mcp-server-test
;; -----------------------------------------------------------------------------
;; BDD coverage for the MCP JSON-RPC layer: initialize handshake, tools/list,
;; tools/call dispatch and a full stdio round-trip.
;; =============================================================================
(ns brazilian-soccer.mcp-server-test
  (:require [clojure.test :refer [deftest testing is]]
            [clojure.data.json :as json]
            [clojure.string :as str]
            [brazilian-soccer.mcp-server :as mcp]
            [brazilian-soccer.test-helper :as h]))

(deftest initialize-test
  (testing "Given an initialize request, When handled, Then capabilities and serverInfo are returned"
    (let [resp (mcp/handle-request (h/graph)
                                   {:id 1 :method "initialize"
                                    :params {:protocolVersion "2024-11-05"}})]
      (is (= 1 (:id resp)))
      (is (= "2024-11-05" (get-in resp [:result :protocolVersion])))
      (is (contains? (get-in resp [:result :capabilities]) :tools))
      (is (= "brazilian-soccer-mcp" (get-in resp [:result :serverInfo :name]))))))

(deftest notifications-are-silent-test
  (testing "Given an initialized notification, When handled, Then no response is produced"
    (is (nil? (mcp/handle-request (h/graph) {:method "notifications/initialized"})))))

(deftest tools-list-test
  (testing "Given tools/list, When handled, Then every advertised tool has a schema"
    (let [resp (mcp/handle-request (h/graph) {:id 2 :method "tools/list"})
          tools (get-in resp [:result :tools])
          names (set (map :name tools))]
      (is (>= (count tools) 8))
      (is (contains? names "find_matches"))
      (is (contains? names "find_players"))
      (is (contains? names "standings"))
      (is (every? :inputSchema tools)))))

(deftest tools-call-find-players-test
  (testing "Given tools/call find_players for Brazilians, When handled, Then text content returns"
    (let [resp (mcp/handle-request
                (h/graph)
                {:id 3 :method "tools/call"
                 :params {:name "find_players"
                          :arguments {:nationality "Brazil" :limit 3}}})
          text (get-in resp [:result :content 0 :text])]
      (is (false? (get-in resp [:result :isError])))
      (is (string? text))
      (is (str/includes? text "Overall")))))

(deftest tools-call-unknown-test
  (testing "Given an unknown tool, When called, Then an error result (not a crash) returns"
    (let [resp (mcp/handle-request (h/graph)
                                   {:id 4 :method "tools/call"
                                    :params {:name "does_not_exist" :arguments {}}})]
      (is (true? (get-in resp [:result :isError]))))))

(deftest unknown-method-test
  (testing "Given an unknown request method, When handled, Then JSON-RPC error -32601"
    (let [resp (mcp/handle-request (h/graph) {:id 5 :method "no/such/method"})]
      (is (= -32601 (get-in resp [:error :code]))))))

(deftest stdio-roundtrip-test
  (testing "Given JSON-RPC lines on stdin, When served, Then JSON-RPC lines come back on stdout"
    (let [input (str (json/write-str {:jsonrpc "2.0" :id 1 :method "initialize" :params {}}) "\n"
                     (json/write-str {:jsonrpc "2.0" :method "notifications/initialized"}) "\n"
                     (json/write-str {:jsonrpc "2.0" :id 2 :method "tools/call"
                                      :params {:name "head_to_head"
                                               :arguments {:team_a "Flamengo" :team_b "Fluminense"}}}) "\n")
          out (java.io.StringWriter.)]
      (mcp/serve (h/graph)
                 (java.io.StringReader. input)
                 out)
      (let [lines (->> (str/split-lines (str out)) (remove str/blank?) (map json/read-str))]
        ;; two requests with ids -> two responses; the notification is silent
        (is (= 2 (count lines)))
        (is (= 1 (get (first lines) "id")))
        (is (contains? (get (first lines) "result") "serverInfo"))
        (is (= 2 (get (second lines) "id")))
        (is (str/includes? (get-in (second lines) ["result" "content" 0 "text"])
                           "head-to-head"))))))
